File: leetcode/countAndSay.js
var countHelper = function(s) {
    let str = "";
    let count = 1;
    let curr = s[0];
    for (let i = 1; i < s.length; i++) {
        if (curr === s[i]) {
            count++;
        } else {
            str += String(count) + curr;
            count = 1;
            curr = s[i];
        }
    }
    str += String(count) + curr;
    return str;
};

var countAndSay = function(n) {
    if (n === 1) return "1";
    return countHelper(countAndSay(n - 1));
};

File: leetcode/countAndSay.java
class Solution {
    public String countAndSay(int n) {
        if (n == 1) return "1";
        String prev = countAndSay(n - 1);
        StringBuilder sb = new StringBuilder();
        char [] chars = prev.toCharArray();
        int count = 0;
        for (int i = 0; i <=chars.length; i++) {
            if (i == chars.length || i >= 1 && chars[i - 1] != chars[i]) {
                sb.append(count);
                sb.append(chars[i - 1]);
                count = 0;
            }
            ++count;
        }
        return sb.toString();
    }
}